reed-solomon-novelpoly/src/field/faster8/f2e16.rs

```diff
@@ -0,0 +1,536 @@
+
+impl BitXor for Additive8x {
+	type Output = Self;
+
+	#[inline(always)]
+	fn bitxor(self, rhs: Self) -> Self::Output {
+		Self(unsafe { _mm_xor_si128(self.0, rhs.0) })
+	}
+}
+
+impl BitXorAssign for Additive8x {
+	fn bitxor_assign(&mut self, rhs: Self) {
+		unsafe {
+			self.0 = _mm_xor_si128(self.0, rhs.0);
+		}
+	}
+}
+
+#[inline(always)]
+pub(crate) fn splat_u16x8(v: u16) -> u16x8 {
+	unsafe { _mm_set1_epi16(v as i16) }
+}
+
+#[inline(always)]
+pub(crate) fn splat_u32x8(v: u32) -> u32x8 {
+	unsafe { _mm256_set1_epi32(v as i32) }
+}
+
+#[allow(dead_code)]
+#[inline(always)]
+pub(crate) fn unpack_u32x8(v: u32x8) -> [u32; 8] {
+	#[repr(C, align(32))]
+	struct Aligned;
+
+	#[repr(C, align(32))]
+	struct Stack {
+		_alignment: [Aligned; 0],
+		data: [u32; 8],
+	}
+
+	let mut dest = Stack { _alignment: [], data: [0u32; 8] };
+
+	unsafe {
+		_mm256_store_si256(core::intrinsics::transmute((&mut dest.data) as *mut [u32; 8]), v);
+	}
+
+	dest.data
+}
+
+#[inline(always)]
+pub(crate) fn unpack_u16x8(v: u16x8) -> [u16; 8] {
+	#[repr(C, align(16))]
+	struct Aligned;
+
+	#[repr(C, align(16))]
+	struct Stack {
+		_alignment: [Aligned; 0],
+		data: [u16; 8],
+	}
+
+	let mut dest = Stack { _alignment: [], data: [0u16; 8] };
+
+	unsafe {
+		_mm_store_si128(core::intrinsics::transmute((&mut dest.data) as *mut [u16; 8]), v);
+	}
+	dest.data
+}
+
+#[inline(always)]
+pub(crate) fn load_u16x8(src: &[u16], idx: [u16; 8]) -> u16x8 {
+	unsafe {
+		_mm_set_epi16(
+			src[idx[0] as usize] as i16,
+			src[idx[1] as usize] as i16,
+			src[idx[2] as usize] as i16,
+			src[idx[3] as usize] as i16,
+			src[idx[4] as usize] as i16,
+			src[idx[5] as usize] as i16,
+			src[idx[6] as usize] as i16,
+			src[idx[7] as usize] as i16,
+		)
+	}
+}
+
+#[inline(always)]
+pub(crate) fn clipping_cast(data: u32x8) -> u16x8 {
+	unsafe {
+		let mask = splat_u32x8(0x0000_FFFF);
+		// need
+		let data = _mm256_and_si256(data, mask);
+		// data_lo zero_lo data_hi zero_hi
+		let packed = _mm256_packus_epi32(data, data);
+
+		// use `a` as above in the lower and higher 128 bits
+		const PERMUTE: i32 = 0b11_01_10_00;
+		let data = _mm256_permute4x64_epi64(packed, PERMUTE);
+
+		_mm256_castsi256_si128(data)
+	}
+}
+
+#[allow(dead_code)]
+#[inline(always)]
+pub(crate) fn expand_cast(data: u16x8) -> u32x8 {
+	unsafe { _mm256_cvtepu16_epi32(data) }
+}
+
+impl Additive8x {
+	pub const LANE: usize = 8;
+
+	#[inline(always)]
+	pub fn zero() -> Self {
+		Self(unsafe { _mm_setzero_si128() })
+	}
+
+	#[cfg(table_bootstrap_complete)]
+	pub fn mul(&self, other: Multiplier) -> Self {
+		let z = Self::zero();
+		if z == *self {
+			return z;
+		}
+		// let log = (LOG_TABLE[self.0 as usize] as Wide) + other.0 as Wide;
+		// let offset = (log & ONEMASK as Wide) + (log >> FIELD_BITS);
+		// Additive(EXP_TABLE[offset as usize])
+
+		unsafe {
+			// spread the multiplier
+			let other32 = splat_u32x8(other.0 as Wide);
+			// unpack_u32x8(other);
+
+			// get the indices
+			let idx = unpack_u16x8(self.0);
```

Comment:
In general, it is better not to mix avx2 and sse instructions where possible. For example, it is better to do all sse and then switch to avx. The reason for that is that it may create false dependencies between avx/sse registers and cause performance loss.

Reply:
And it sometimes useful to add `vzeroupper` when you do such a switch.

Reply:
I tried to stick closely to the original code structure and leave the reordering to the compiler.